Documentation/design-docs/host-testing.md

```diff
@@ -0,0 +1,100 @@
+* Having the app output information to standard output
+* Turning on host tracing and looking for certain text in the tracing output
+* Exit code of the app
+
+Pros:
+* True End-to-End tests which test real product binaries in the real-world conditions (mostly)
+* Tests are written in C# - so natural integration with test infra, VS UI and so on
+Cons:
+* All tests are out-of-proc which makes debugging product complicated
+* Tests perform large setup work and copy/write lot of files on disk - tests are slow for this reason.
+* Tests run the entire product, not just the host. This includes starting the runtime, JITing and so on. Vast majority of tests don't actually use that and effectively ignore the part of running an actual app code - makes the tests slow.
+
+Going forward we would keep these tests and add new ones to provide true End-to-End coverage.
+
+### Native API tests
+Small portion of the tests actually call specific exports on `hostfxr` or `hostpolicy` directly to test these. All these tests are in the `HostActivation` project under the `GivenThatICareAboutNativeHostApi` test class.
+These tests use a special test application (managed) which invokes the selected exports through PInvokes and performs the testing. The `HostActivation` test only prepares this app and executes it, looking for pieces of its output to verify the outcome.
+Ideally we would migrate these over time to the proposed Component tests infra.
+
+Pros:
+* Testing real product binaries
+* Tests are written in C#, but in a separate project - no direct test infra integration or VS UI support. This is worked around by effectively having "wrappers" for these tests.
+* Direct calls to exports in a controlled environment without running the whole product (mostly)
+Cons:
+* Still runs out-of-proc making debugging harder
+
+## Test proposal
+
+### Unit tests
+Add the ability to write true unit tests. So tests written in C++ which can directly call methods/classes from the product. These tests would be compiled as native executables. For simplicity of the build system, they would live in the product source tree (`src` folder) and would be compiled during the product build. They would combine the test source files with product source files into one executable.
+For integration purposes there would be managed wrappers in the `HostActivation` test project which would just call the native executable and check the exit code. As this is the simplest way to integrate these tests into all our test infrastructure.
+
+Unit tests should be used to test self-contained bits of functionality. So for example helper classes and such.
+
+First unit test is being introduced in this PR: dotnet/core-setup#4953.
+
+Pros:
+* Can test C++ code directly
+* Can create very isolated environment for the tested code
+
+Cons:
+* Test must be written in C++
+* Not testing shipping binaries
+* Debugging is trickier as it launches a separate process (this could be fixed)
```

Comment:
These apps can be run standalone.  On Linux this make debugging easier.

Reply:
Good point - fixed